### PR TITLE
[FEAT] Process multiple partitions in parallel

### PR DESCRIPTION
**Pull request by elephaint**
Draft PR for #288

@AzulGarza I've a working version of the `ThreadPoolExecutor` variant, but the `mp.Pool` version is a bit more tricky; due to the use of the wrapper I can't find an elegant solution yet to make it work (mp.Pool does not like these nested functions that inherit from a class, and I haven't found a way around yet). I'll try a bit more.

One question I already had - why not do this asynchronously? I'm def not an expert in this but I thought asynchronous calls for API requests can typically be quite fast. No idea yet on how to implement though.

Todo:
- mp.Pool version
- Benchmark between `mp.Pool` and `ThreadPoolExecutor` version

### CODE REVIEW

**Comment by elephaint** [comment]
Results from the ThreadPoolExecutor version show there is an improvement when the num_partitions increases.....

num_partitions | current | ThreadPoolExecutor
-- | -- | --
1 | 25.26 | 25.79
2 | 31.53 | 26.16
4 | 39.00 | 19.94

test script:

```
#%% Import packages
import numpy as np
import pandas as pd
from nixtlats import NixtlaClient
from neuralforecast.utils import generate_series
from time import perf_counter
#%% Create Client
nixtla_client = NixtlaClient(
    api_key = [USE YOUR KEY]
)
nixtla_client.validate_api_key()
#%% Generate big dataset
n_series = 2000
df, S_df = generate_series(n_series=n_series, 
                             n_temporal_features=0, 
                             n_static_features=2,
                             seed=0)

#%% Run performance test of partition_by_uid
# Inputs
n_calls = 1
num_partitions_dict = {
                       1: 0,
                       2: 0,
                       4: 0,
                    #    8: 0,
                         }

model = 'timegpt-1-long-horizon'
# model = 'timegpt-1'
horizon = 14
freq = 'D'

# Base forecasts to test output equivalence
base_forecasts = nixtla_client.forecast(df=df, 
                h=horizon, 
                id_col = "unique_id",
                time_col = "ds",
                target_col = "y",
                freq=freq, 
                model = model,
                num_partitions=1)

# Test loop
for num_partitions, timing in num_partitions_dict.items():
    for _ in range(n_calls):
        # Time the calls with varying num_partitions
        start = perf_counter()
        timegpt_fcst_df = nixtla_client.forecast(df=df, 
                                                 h=horizon, 
                                                 id_col = "unique_id",
                                                 time_col = "ds",
                                                 target_col = "y",
                                                 freq=freq, 
                                                 model = model,
                                                 num_partitions=num_partitions)
        end = perf_counter()
        timing = timing + (1 / n_calls) * (end - start)
        # Assert output equivalence to base case
        np.testing.assert_allclose(timegpt_fcst_df["TimeGPT"], base_forecasts["TimeGPT"], atol=1e-6, rtol=1e-6)
    
    num_partitions_dict[num_partitions] = timing

# Print timing
for num_partitions, timing in num_partitions_dict.items():
    print(f"Timing num_partitions {num_partitions}: {timing:.2f}s")
```

However, still no luck with the mp.Pool version. 

It should look something like the below, but implementing this keeps giving a threadlock error, even when implementing it within the Nixtla class (i.e. not as a decorator), even when rewriting it as a dedicated (e.g.) forecast function, and even when creating separate standalone Nixtla clients for each subprocess - the issue from the threadlock keeps coming from the Nixtla client (I tested it without accessing the Nixtla client to forecast, and then I can get the parallelization to work).

Any thoughts?

```
def partition_by_uid_single(uids_split, func, df, X_df, id_col, **kwargs):
    df_uids = df.query("unique_id in @uids_split")
    if X_df is not None:
        X_df_uids = X_df.query("unique_id in @uids_split")
    else:
        X_df_uids = None
    df_uids = remove_unused_categories(df_uids, col=id_col)
    X_df_uids = remove_unused_categories(X_df_uids, col=id_col)
    kwargs_uids = {"df": df_uids, **kwargs}
    if X_df_uids is not None:
        kwargs_uids["X_df"] = X_df_uids

    kwargs_uids["id_col"] = id_col
    results_uids = func(**kwargs_uids, num_partitions=1)

    return results_uids


def partition_by_uid(func):
    def wrapper(self, num_partitions, **kwargs):

        if num_partitions is None or num_partitions == 1:
            return func(self, **kwargs, num_partitions=1)

        df = kwargs.pop("df")
        X_df = kwargs.pop("X_df", None)
        id_col = kwargs.pop("id_col")
        uids = df["unique_id"].unique()

        partial_partition_by_uid_single = partial(
            partition_by_uid_single,
            self=self,
            func=func,
            df=df,
            X_df=X_df,
            id_col=id_col,
            **kwargs
        )

        num_processes = min(num_partitions, cpu_count())
        uids_splits = np.array_split(uids, num_processes)
        pool = Pool(num_processes)
        results = pool.map(partial_partition_by_uid_single, uids_splits)
        pool.close()
        pool.join()

        results_df = pd.concat(results).reset_index(drop=True)

        return results_df

    return wrapper
```

**Comment by elephaint** [comment]
I've made an asyncio version but that one is also quite slow; it seems to execute sequentially rather than concurrently. Below is an example of this version. Anyways, I vote to just use the ThreadPoolExecutor variant, simple and offers some benefit beyond the naive for-loop. Wdyt?

Forecast function inside _NixtlaClient:
```
    @validate_model_parameter
    def _forecast(
        self,
        df: pd.DataFrame,
        h: int,
        freq: Optional[str] = None,
        id_col: str = "unique_id",
        time_col: str = "ds",
        target_col: str = "y",
        X_df: Optional[pd.DataFrame] = None,
        level: Optional[List[Union[int, float]]] = None,
        quantiles: Optional[List[float]] = None,
        finetune_steps: int = 0,
        finetune_loss: str = "default",
        clean_ex_first: bool = True,
        validate_api_key: bool = False,
        add_history: bool = False,
        date_features: Union[bool, List[str]] = False,
        date_features_to_one_hot: Union[bool, List[str]] = True,
        model: str = "timegpt-1",
        num_partitions: int = 1,
    ):

        if validate_api_key and not self.validate_api_key(log=False):
            raise Exception("API Key not valid, please email ops@nixtla.io")

        loop = asyncio.get_event_loop()
        if loop.is_running():
            raise Exception(
                "Can't use this function when there's already a running loop. "
                "Use `await async_download_files(...) instead.`"
            )
             
        partial_nixtla_client_model = partial(_NixtlaClientModel, 
                                                client=self.client,
                                                h=h,
                                                id_col=id_col,
                                                time_col=time_col,
                                                target_col=target_col,
                                                freq=freq,
                                                level=level,
                                                quantiles=quantiles,
                                                finetune_steps=finetune_steps,
                                                finetune_loss=finetune_loss,
                                                clean_ex_first=clean_ex_first,
                                                date_features=date_features,
                                                date_features_to_one_hot=date_features_to_one_hot,
                                                model=model,
                                                max_retries=self.max_retries,
                                                retry_interval=self.retry_interval,
                                                max_wait_time=self.max_wait_time,
                                                )        

        uids = df['unique_id'].unique()
        tasks = []
        for uids_split in np.array_split(uids, num_partitions):
            df_uids = df.query('unique_id in @uids_split')
            if X_df is not None:
                X_df_uids = X_df.query('unique_id in @uids_split')
            else:
                X_df_uids = None                
            df_uids = remove_unused_categories(df_uids, col=id_col)
            X_df_uids = remove_unused_categories(X_df_uids, col=id_col)
            tasks.append(_async_forecast_partition(nixtla_client_model=partial_nixtla_client_model(), 
                                                    df_uids = df_uids, 
                                                    X_df_uids=X_df_uids,
                                                    add_history=add_history))

        fcst_dfs, _ = loop.run_until_complete(asyncio.wait(tasks))

        fcst_df = pd.concat([df.result() for df in fcst_dfs]).reset_index(drop=True)    

        return fcst_df 
```

and as a standalone function:

```
async def _async_forecast_partition(nixtla_client_model: Nixtla, df_uids: pd.DataFrame,
                                    X_df_uids: pd.DataFrame, add_history: bool):

    fcst_df = nixtla_client_model.forecast(
        df=df_uids, X_df=X_df_uids, add_history=add_history
    )

    return fcst_df
```

**Comment by AzulGarza** [comment]
hey @elephaint! awesome work so far! let's use `ThreadPoolExecutor`. once it is done, we will be ready to merge it.

**Comment by elephaint** [comment]
> hey @elephaint! awesome work so far! let's use `ThreadPoolExecutor`. once it is done, we will be ready to merge it.

Great - I think it is done (i.e. not sure what else to add to it?) . I'm not sure why the tests are failing, because they fail also if I revert all my changes and just execute the current main branch (?)

File: nixtlats/nixtla_client.py [hunk 1]
```diff
@@ -745,16 +751,17 @@ def remove_unused_categories(df: pd.DataFrame, col: str):
...
-        id_col = kwargs["id_col"]
+        id_col = kwargs.pop("id_col")
         uids = df["unique_id"].unique()
-        results_df = []
-        for uids_split in np.array_split(uids, num_partitions):
+
+        def partition_by_uid_single(uids_split, func, df, X_df, id_col, **kwargs):
             df_uids = df.query("unique_id in @uids_split")
             if X_df is not None:
                 X_df_uids = X_df.query("unique_id in @uids_split")
```
Comment (AzulGarza) [inline nixtlats/nixtla_client.py:767]: could we add locks to ensure that only one thread at a time can access `df` and `X_df`?

```python
from threading import Lock
```

inside `wrapper`:

```python
df_lock = Lock()
X_df_lock = Lock()

def partition_by_uid_single(uids_split, func, df, X_df, id_col, **kwargs):
         with df_lock:
               df_uids = df.query("unique_id in @uids_split")
          if X_df is not None:
              with X_df_lock:
                   X_df_uids = X_df.query("unique_id in @uids_split")

```

**Review by AzulGarza — APPROVED** [review]
hey @elephaint! i left a small comment.

also, please open the PR from a branch inside the repo (not a fork). to prevent unwanted usage, ci/cd cannot use secret variables when changes come from a fork.

after those changes, we will be ready to merge it, thank you!

💙

**Review by AzulGarza — CHANGES_REQUESTED** [review]
approved it by error xD

File: nixtlats/nixtla_client.py (hunk 1 above)
Comment (elephaint) [reply nixtlats/nixtla_client.py:767]: Sure!

**Pull request closed without merging**